adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,712 @@
+    ret = bind(sock, (struct sockaddr*)&sock_addr, sizeof(sock_addr));
+    if (ret) {
+        LogError("bind socket failed\n");
+        return;
+    }
+    LogInfo("bind socket OK\n");
+
+    LogInfo("socket connect to remote ......");
+    memset(&sock_addr, 0, sizeof(sock_addr));
+    sock_addr.sin_family = AF_INET;
+    sock_addr.sin_addr.s_addr = target_ip.addr;
+    sock_addr.sin_port = htons(tls_io_instance->port);
+
+    lwip_set_non_block(sock);
+
+    ret = connect(sock, (struct sockaddr*)&sock_addr, sizeof(sock_addr));
+    if (ret == -1) {
+        ret = lwip_net_errno(sock);
+        if (ret != EINPROGRESS){
+            LogError("socket connect failed %s\n", tls_io_instance->hostname);
+            return;
+        } else{
+            LogInfo("socket connect in progress\n");
+        }
+    }
+    LogInfo("socket connect OK\n");
+
+    char recv_buf[128];
+    int recv_bytes = (int)sizeof(recv_buf);
+    int retry = 0;
+
+    while (retry < 10){
+        LogInfo("lwip select retry #: %d\n", retry);
+
+        FD_ZERO(&readset);
+        FD_SET(sock, &readset);
+    
+        FD_ZERO(&writeset);
+        FD_SET(sock, &writeset);
+    
+        FD_ZERO(&errset);
+        FD_SET(sock, &errset);
+    
+        ret = lwip_select(sock + 1, &readset, &writeset, &errset, NULL);
+        if (ret > 0){
+            if (FD_ISSET(sock, &writeset)){
+              break;
+            }
+    
+            if (FD_ISSET(sock, &readset)){
+                memset(recv_buf, 0, recv_bytes);
+                break;
+            }
+        }else{
+            LogInfo("lwip_select returned: %d\n", ret);
+        }
+
+        retry++;
+    }
+
+    int s = sock;
+    ctx = SSL_CTX_new(TLSv1_client_method());
+    if (!ctx) {
+        LogError("create new SSL CTX failed\n");
+        return;
+    }
+    LogInfo("create new SSL CTX OK\n");
+    
+    LogInfo("SSL set fragment\n");
+    ret = SSL_set_fragment(ctx, OPENSSL_FRAGMENT_SIZE);
+
+    LogInfo("SSL new... \n");
+    ssl = SSL_new(ctx);
+
+    LogInfo("SSL set fd\n");
+    SSL_set_fd(ssl, s);
+
+    LogInfo("SSL connect... \n");
+
+    while (-1 == SSL_connect(ssl))
+    {  
+        FD_ZERO(&readset);
+        FD_SET(s, &readset);
+        FD_ZERO(&writeset);
+        FD_SET(s, &writeset);
+        FD_ZERO(&errset);
+        FD_SET(s, &errset);
+
+        ret = lwip_select(s + 1, &readset, &writeset, &errset, NULL);
+    }
+
+    LogInfo("SSL connect ok\n");
+
+    tls_io_instance->ssl = ssl;
+    tls_io_instance->ssl_context = ctx;
+}
+
+static int send_handshake_bytes(TLS_IO_INSTANCE* tls_io_instance)
+{
+    //system_print_meminfo(); // This is useful for debugging purpose.
+    //LogInfo("free heap size %d", system_get_free_heap_size()); // This is useful for debugging purpose.
+    openssl_thread_LWIP_CONNECTION(tls_io_instance);
+    tls_io_instance->tlsio_state = TLSIO_STATE_OPEN;
+    indicate_open_complete(tls_io_instance, IO_OPEN_OK);    
+    int result = 0;
```

Comment:
Please check if the calls on this function had success before return success or fail.